Support reading gzip files

NickLaMuro opened the pull request:
Built off of https://github.com/Fryguy/memory_analyzer/pull/4

When comparing sizes of a memory dump file when compressed and not:

```console
$ ls -lh tmp/sample_dumps
total 1256264
-rw-r--r--  1 user  staff   576M Jan 01 19:08 dump
-rw-r--r--  1 user  staff    37M Jan 01 19:09 dump.gz
```

The disk space requirements to keep around a gzipped file are far more beneficial instead of it's uncompressed counterpart, and the time difference and memory requirements to parse a gzip file versus a uncompressed file are next to none.

```console
$ time ruby bin/memory_analyzer tmp/sample_dumps/dump.gz
Parsing: |========================================| Time: 00:03:11
Mem: 3695.37890625MiB

real    3m14.405s
user    3m11.839s
sys     0m2.357s
$ time ruby bin/memory_analyzer tmp/sample_dumps/dump
Parsing: |========================================| Time: 00:03:11
Mem: 3625.39453125MiB

real    3m13.220s
user    3m10.516s
sys     0m2.522s
```

Changes
-------

This PR adds the `#io_klass` which will look at the filename's extension to determine if it is a gzipped file (probably a better way to do this), and then uses chooses the proper IO class to read the file with.  The `ProgressBar` `:total` option also makes use of this method.

If the file is a gzip file, it will require the zlib library and use the `Zlib::GzipReader` class to parse the file.

The `#parse_file` method needed to be modified slightly to use a common API between the `Zlib::GzipReader` and `File` classes, but changes to the method functionally equivalent to the old implementation.

NickLaMuro committed:
Count line numbers using ruby's IO, not shellout

Makes the progressbar code more portable since it doesn't rely on shell
cmds being present on the system (e.g. Windows users).  Roughly the same
performance, and any IO object that responds to `#each_line` can
technically be swapped in here to do the same (SPOILERS: will happen in
a future commit)

NickLaMuro committed:
Support gzipped heap dump files

When comparing sizes of a memory dump file when compressed and not:

    $ ls -lh tmp/sample_dumps
    total 1256264
    -rw-r--r--  1 user  staff   576M Jan 01 19:08 dump
    -rw-r--r--  1 user  staff    37M Jan 01 19:09 dump.gz

The disk space requirements to keep around a gzipped file are far more
beneficial instead of it's uncompressed counterpart, and the time
difference and memory requirements to parse a gzip file versus a
uncompressed file are next to none.

    $ time ruby bin/memory_analyzer tmp/sample_dumps/dump.gz
    Parsing: |========================================| Time: 00:03:11
    Mem: 3695.37890625MiB

    real    3m14.405s
    user    3m11.839s
    sys     0m2.357s
    $ time ruby bin/memory_analyzer tmp/sample_dumps/dump
    Parsing: |========================================| Time: 00:03:11
    Mem: 3625.39453125MiB

    real    3m13.220s
    user    3m10.516s
    sys     0m2.522s

Adds the `#io_class` which will look at the filename's extension to
determine if it is a gzipped file (probably a better way to do this),
and then uses chooses the proper IO class to read the file with.  The
`ProgressBar` `:total` option also makes use of this method.

If the file is a gzip file, it will require the zlib library and use the
`Zlib::GzipReader` class to parse the file.

The `#parse_file` method needed to be modified slightly to use a common
API between the `Zlib::GzipReader` and `File` classes, but changes to
the method functionally equivalent to the old implementation.

Fryguy commented on lib/memory_analyzer/heap_analyzer/parser.rb:
```diff
@@ -93,6 +95,15 @@ def string_pool
         string_pool.add(string)[string]
       end
+
+      def io_klass
```
might as well call it `io_class`

NickLaMuro replied:
psh... all the cool kids spell klass with a K...

kidding, fair enough

Fryguy commented on lib/memory_analyzer/heap_analyzer/parser.rb:
```diff
@@ -93,6 +95,15 @@ def string_pool
+      def io_klass
+        if File.extname(file) == ".gz"
+          require 'zlib'
+          Zlib::GzipReader
```
I think there's a way to read the magic number of the file and use that...let me find out.

NickLaMuro replied:
I can look as well, I was just going with this for simplicity, but I think we should probably do both.

Fryguy replied:
Yeah, do the following:

```ruby
GZ_MAGIC_NUMBER = "\x1F\x8B".force_encoding("ASCII-8BIT").freeze

if File.read(file, 2, :mode => "rb") == GZ_MAGIC_NUMBER
  require 'zlib'
  Zlib::GzipReader
else
  File
end
```

NickLaMuro replied:
@Fryguy Does it make sense to do both?

```ruby
if File.read(file, 2, :mode => "rb") == GZ_MAGIC_NUMBER || File.extname(file) == ".gz"
```

Also, isn't there another character of `08` to that magic number?

https://www.garykessler.net/library/file_sigs.html

Fryguy replied:
can't hurt to do both.

Fryguy replied:
The 8 is the compression method (8 == deflate), so isn't really part of the magic number:

http://www.zlib.org/rfc-gzip.html#header-trailer

NickLaMuro replied:
Looks like `Zlib` actually does some of this already:

https://github.com/ruby/ruby/blob/de5e347/ext/zlib/zlib.c#L2135-L2136

So might be able to make use of that.  Also, that further proves your point about the `08` being not necessary, thanks.

NickLaMuro committed:
Read magic number on file to check if gzip format

An extra test for determining io_class to make sure we are reading the
proper format.
